### DISCUSSION TITLE
Any plan to implement Hair Shape (from Mitsuba 0.6) ?

### DISCUSSION THREAD

**Discussion by sebastienleon** [mitsuba-renderer/mitsuba3#582] [Feature requests] [answered]
Hi,
If I am not wrong, the Hair shape is no longer supported in latest version.
From XML, it is this tag :
<shape type="hair">...>/shape>
In Mitsuba doc, in "Missing features", once can read :
_the basic shapes are all supported. However, instancing and assemblies of hair fibers are still missing._
Latest code seems to be there :
https://github.com/mitsuba-renderer/mitsuba/blob/master/src/shapes/hair.cpp
Any plan to put it back ? Or a new logic is now used and this feature will not come back ?
Best regards
Sebastien

**Comment by njroussel** [comment]
Hello @sebastienleon 

Your timing is _almost_ perfect. There is a B-spline and linear curve shape that is currently being implemented. In fact, you can checkout the `curves` branch. There is a working implementation for both CPU and GPU. There's still some cleanup and documentation that needs to be done, I expect it to be finished next week.

**Comment by sebastienleon** [comment]
Hi,
Thanks for your answer. All of that sounds really good !
I will try this branch !

**Comment by apoorvkhattar** [comment]
A few bugs that I found in B-spline curve shape. 

1. The tangent and binormal for `si.sh_frame` appear to be swapped so a simple fix is to interchange them. I found this out by comparing the value for the output of the B-spline straight line with points in a straight line with the output of the cylinder shape.

2. Similary, the UV-value should match that of a cylinder. The U and V need to be swapped as V-value should essentially represent the length for a given hit point. Additionally the U value (phase) becomes inconsistent across two cylinders. This can be fixed by recalculating the U as 
`   uv = mi.Point2f(si.uv);
    si.uv.x = uv.y;
    si.uv.y = uv.x;
    si.uv.x = dr.atan(dr.tan(si.uv.x * dr.two_pi));
    si.uv.x = dr.select(si.uv.x < 0, (si.uv.x + dr.two_pi) / dr.two_pi, si.uv.x / dr.two_pi); `
I applied this fix in my python integrator but it can be done in the C++ shape code as well to fix it. Attached are the screenshots of the U value before and after this fix.

3. Some noise present in the U values. I have not figured out why this happens as it is seen only when the control points create a straight cylinder along the Z-axis.
![Screenshot from 2023-03-30 12-08-38](https://user-images.githubusercontent.com/22664823/228818021-b7aa678e-75cf-4131-b152-51f91bf55635.png)
![Screenshot from 2023-03-30 12-08-25](https://user-images.githubusercontent.com/22664823/228818028-0961ac70-85ed-4fb0-a5df-992a7ee311f4.png)

**Reply by njroussel** [reply]
Hi @apoorvkhattar 

There is still "major" work being done on this branch. The release has been delayed a bit. 

Your first two comments make sense, I will change that. Thank you for pointing it out.

For you last comment, was this rendered on the GPU? We've already discovered some numerical stability issues with OptiX and its B-Spline curves.

**Reply by apoorvkhattar** [reply]
No all the tests I did were on CPU using LLVM as backend.

**Reply by njroussel** [reply]
The version in the release has fixed these issues.
Thank you for pointing them out!

**Comment by apoorvkhattar** [comment]
So the solution I suggested also does not work if the curve has high curvature. Also in the attached images before it can be seen that the U is not continuous since the back-part of the cylinder is not continuous. ![Inked228818021-b7aa678e-75cf-4131-b152-51f91bf55635](https://user-images.githubusercontent.com/22664823/229767463-40776c60-4194-4de3-a19f-339e9e11bbfd.jpg)

The problem is that the `local_frame()` function which I think is defined based on the idea of the Frenet Frame produces inconsistent reference normals for `v_rot`. I was able to add Rotation Minimization Frame (RMF) normals as an alternative. So the input txt file instead of having 4 entries now takes 7 entries for each control points where the first three indices are the points, followed by three indices for the RMF normals and finally the radius. For a given hit point the reference normal can be determined by simple linear interpolation between the RMF normals of the two vertices for given segment with index `i` i.e. vertex at index `i` and index `i+1`. Now the U value is consistent throughout the curve. Attached is the update U value for helix. 
![Screenshot from 2023-04-04 11-17-27](https://user-images.githubusercontent.com/22664823/229769822-77ed4896-ebaf-4590-8f17-d876b20d6eae.png)

One issue with this implementation I am having right now is that sometimes the renderer gives an empty scene as outputs but for another random run it works fine. What could be the issue?

**Comment by njroussel** [comment, marked as answer]
The latest [release](https://github.com/mitsuba-renderer/mitsuba3/releases/tag/v3.3.0) (v3.3.0) has added support for a `bsplinecurve` shape which is very simialr to the `hair` shape from Mitsuba 0.6.
